File: vbdotnet/2048.vb
Imports System

Public Module Class2048
    Private rand As Random = New Random()
    Const ROW_MASK As ULong = &HFFFFUL
    Const COL_MASK As ULong = &HF000F000F000FUL
    Const TABLESIZE As Integer = 65536
    Private row_table As UShort() = New UShort(TABLESIZE - 1) {}
    Private score_table As UInteger() = New UInteger(TABLESIZE - 1) {}
    Delegate Function get_move_func_t(board As ULong) As Integer
    Const UP As Integer = 0
    Const DOWN As Integer = 1
    Const LEFT As Integer = 2
    Const RIGHT As Integer = 3
    Const RETRACT As Integer = 4

    Private Function unif_random(n As Integer) As Integer
        Return rand.Next(n)
    End Function

    Private Sub clear_screen()
        Console.Clear()
    End Sub

    Private Function get_ch() As Char
        Return Console.ReadKey(True).KeyChar
    End Function

    Private Function unpack_col(row As UShort) As ULong
        Dim tmp As ULong = row
        Return (tmp Or (tmp << 12) Or (tmp << 24) Or (tmp << 36)) And COL_MASK
    End Function

    Private Function reverse_row(row As UShort) As UShort
        Return (row >> 12) Or ((row >> 4) And &HF0) Or ((row << 4) And &HF00) Or (row << 12)
    End Function

    Private Sub print_board(board As ULong)
        Console.WriteLine("-----------------------------")
        For i As Integer = 0 To 3
            For j As Integer = 0 To 3
                Dim power_val As Integer = board And &HFUL
                If power_val = 0 Then
                    Console.Write("|{0,6}", " "c)
                Else
                    Console.Write("|{0,6}", 1 << power_val)
                End If
                board >>= 4
            Next
            Console.WriteLine("|")
        Next
        Console.WriteLine("-----------------------------")
    End Sub

    Private Function transpose(x As ULong) As ULong
        Dim a1 As ULong = x And &HF0F00F0FF0F00F0FUL
        Dim a2 As ULong = x And &HF0F00000F0F0UL
        Dim a3 As ULong = x And &HF0F00000F0F0000UL
        Dim a As ULong = a1 Or (a2 << 12) Or (a3 >> 12)
        Dim b1 As ULong = a And &HFF00FF0000FF00FFUL
        Dim b2 As ULong = a And &HFF00FF00000000UL
        Dim b3 As ULong = a And &HFF00FF00UL
        Return (b1 Or (b2 >> 24) Or (b3 << 24))
    End Function

    Private Function count_empty(x As ULong) As Integer
        x = x Or ((x >> 2) And &H3333333333333333UL)
        x = x Or (x >> 1)
        x = (Not x) And &H1111111111111111UL
        x += x >> 32
        x += x >> 16
        x += x >> 8
        x += x >> 4
        Return x And &HF
    End Function

    Private Sub init_tables()
        Dim row As UShort = 0
        Dim result As UShort = 0
        Dim line As UShort() = New UShort(3) {}

        Do
            Dim i As Integer = 0, j As Integer = 0
            Dim score As UInteger = 0
            line(0) = row And &HF
            line(1) = (row >> 4) And &HF
            line(2) = (row >> 8) And &HF
            line(3) = (row >> 12) And &HF

            While i < 4
                Dim rank As Integer = line(i)
                If rank >= 2 Then
                    score += (rank - 1) * (1 << rank)
                End If
                i += 1
            End While
            score_table(row) = score

            i = 0
            While i < 3
                j = i + 1
                While j < 4
                    If line(j) <> 0 Then Exit While
                    j += 1
                End While
                If j = 4 Then Exit While

                If line(i) = 0 Then
                    line(i) = line(j)
                    line(j) = 0
                    i -= 1
                ElseIf line(i) = line(j) Then
                    If line(i) <> &HF Then
                        line(i) += 1
                    End If
                    line(j) = 0
                End If
                i += 1
            End While
            result = line(0) Or (line(1) << 4) Or (line(2) << 8) Or (line(3) << 12)
            row_table(row) = row Xor result

            row += 1
        Loop While row <> (TABLESIZE - 1)
    End Sub

    Private Function execute_move_col(board As ULong, _move As Integer) As ULong
        Dim ret As ULong = board
        Dim t As ULong = transpose(board)
        If _move = UP Then
            ret = ret Xor unpack_col(row_table(t And ROW_MASK))
            ret = ret Xor (unpack_col(row_table((t >> 16) And ROW_MASK)) << 4)
            ret = ret Xor (unpack_col(row_table((t >> 32) And ROW_MASK)) << 8)
            ret = ret Xor (unpack_col(row_table((t >> 48) And ROW_MASK)) << 12)
        ElseIf _move = DOWN Then
            ret = ret Xor unpack_col(reverse_row(row_table(reverse_row(t And ROW_MASK))))
            ret = ret Xor (unpack_col(reverse_row(row_table(reverse_row((t >> 16) And ROW_MASK)))) << 4)
            ret = ret Xor (unpack_col(reverse_row(row_table(reverse_row((t >> 32) And ROW_MASK)))) << 8)
            ret = ret Xor (unpack_col(reverse_row(row_table(reverse_row((t >> 48) And ROW_MASK)))) << 12)
        End If
        Return ret
    End Function

    Private Function execute_move_row(board As ULong, _move As Integer ) As ULong
        Dim ret As ULong = board
        If _move = LEFT Then
            ret = ret Xor CULng(row_table(board And ROW_MASK))
            ret = ret Xor (CULng(row_table((board >> 16) And ROW_MASK)) << 16)
            ret = ret Xor (CULng(row_table((board >> 32) And ROW_MASK)) << 32)
            ret = ret Xor (CULng(row_table((board >> 48) And ROW_MASK)) << 48)
        ElseIf _move = RIGHT Then
            ret = ret Xor CULng(reverse_row(row_table(reverse_row(board And ROW_MASK))))
            ret = ret Xor (CULng(reverse_row(row_table(reverse_row((board >> 16) And ROW_MASK)))) << 16)
            ret = ret Xor (CULng(reverse_row(row_table(reverse_row((board >> 32) And ROW_MASK)))) << 32)
            ret = ret Xor (CULng(reverse_row(row_table(reverse_row((board >> 48) And ROW_MASK)))) << 48)
        End If
        Return ret
    End Function

    Private Function execute_move(_move As Integer, board As ULong) As ULong
        If _move = UP OrElse _move = DOWN Then
            Return execute_move_col(board, _move)
        ElseIf _move = LEFT OrElse _move = RIGHT Then
            Return execute_move_row(board, _move)
        Else
            Return &HFFFFFFFFFFFFFFFFUL
        End If
    End Function

    Private Function score_helper(board As ULong) As UInteger
        Return score_table(board And ROW_MASK) + score_table(board >> 16 And ROW_MASK) + score_table(board >> 32 And ROW_MASK) + score_table(board >> 48 And ROW_MASK)
    End Function

    Private Function score_board(board As ULong) As UInteger
        Return score_helper(board)
    End Function

    Private Function ask_for_move(board As ULong) As Integer
        print_board(board)
        While True
            Const allmoves As String = "wsadkjhl"
            Dim pos As Integer = 0
            Dim movechar As Char = get_ch()

            If movechar = "q"c Then
                Return -1
            ElseIf movechar = "r"c Then
                Return RETRACT
            End If
            pos = allmoves.IndexOf(movechar)
            If pos <> -1 Then
                Return pos Mod 4
            End If
        End While
    End Function

    Private Function draw_tile() As ULong
        If unif_random(10) < 9 Then
            Return 1
        Else
            Return 2
        End If
    End Function

    Private Function insert_tile_rand(board As ULong, tile As ULong) As ULong
        Dim index As Integer = unif_random(count_empty(board))
        Dim tmp As ULong = board

        While True
            While (tmp And &HFUL) <> 0
                tmp >>= 4
                tile <<= 4
            End While
            If index = 0 Then Exit While
            index -= 1
            tmp >>= 4
            tile <<= 4
        End While
        Return board Or tile
    End Function

    Private Function initial_board() As ULong
        Dim board As ULong = CULng((draw_tile())) << (unif_random(16) << 2)
        Return insert_tile_rand(board, draw_tile())
    End Function

    Private Sub play_game(get_move As get_move_func_t)
        Dim board As ULong = initial_board()
        Dim scorepenalty As Integer = 0
        Dim last_score As Integer = 0, current_score As Integer = 0, moveno As Integer = 0
        Const MAX_RETRACT As Integer = 64
        Dim retract_vec As ULong() = New ULong(MAX_RETRACT - 1) {}
        Dim retract_penalty_vec As Byte() = New Byte(MAX_RETRACT - 1) {}
        Dim retract_pos As Integer = 0, retract_num As Integer = 0

        While True
            Dim _move As Integer = 0
            Dim tile As ULong = 0
            Dim newboard As ULong
            clear_screen()

            For _move = 0 To 3
                If execute_move(_move, board) <> board Then Exit For
            Next

            If _move = 4 Then Exit While
            current_score = CInt(score_board(board) - scorepenalty)
            moveno += 1
            Console.WriteLine("Move #{0}, current score={1}(+{2})", moveno, current_score, current_score - last_score)
            last_score = current_score
            _move = get_move(board)
            If _move < 0 Then Exit While

            If _move = RETRACT Then
                If moveno <= 1 OrElse retract_num <= 0 Then
                    moveno -= 1
                    Continue While
                End If
                moveno -= 2
                If retract_pos = 0 AndAlso retract_num > 0 Then retract_pos = MAX_RETRACT
                retract_pos -= 1
                board = retract_vec(retract_pos)
                scorepenalty -= retract_penalty_vec(retract_pos)
                retract_num -= 1
                Continue While
            End If

            newboard = execute_move(_move, board)
            If newboard = board Then
                moveno -= 1
                Continue While
            End If

            tile = draw_tile()
            If tile = 2 Then
                scorepenalty += 4
                retract_penalty_vec(retract_pos) = 4
            Else
                retract_penalty_vec(retract_pos) = 0
            End If
            retract_vec(retract_pos) = board
            retract_pos += 1
            If retract_pos = MAX_RETRACT Then retract_pos = 0
            If retract_num < MAX_RETRACT Then retract_num += 1
            board = insert_tile_rand(newboard, tile)
        End While

        print_board(board)
        Console.WriteLine("Game over. Your score is {0}.", current_score)
    End Sub

    Sub Main(args As String())
        init_tables()
        play_game(New get_move_func_t(AddressOf ask_for_move))
    End Sub
End Module

File: fortran/2048.f
      PROGRAM GAME2048

      CALL MAIN()
      STOP
      END

      SUBROUTINE INIT_NUM()
        INTEGER*8 :: ZF0F00F0F
        INTEGER*8 :: ZFF00FF00
        INTEGER*8 :: T8_1, T8_2
        INTEGER*8 :: ROW_MASK, COL_MASK
        COMMON /MASK_NUM/ ROW_MASK, COL_MASK
        INTEGER*2 :: Z000F, Z00F0, Z0F00
        COMMON /CONST_NUM/  Z000F, Z00F0, Z0F00
        INTEGER*8 :: ZF0F00F0FF0F00F0F, Z0000F0F00000F0F0,
     & Z0F0F00000F0F0000, ZFF00FF0000FF00FF
        INTEGER*8 :: Z00FF00FF00000000, Z00000000FF00FF00,
     & Z3333333333333333, Z1111111111111111
        COMMON /LARGE_NUM1/ ZF0F00F0FF0F00F0F, Z0000F0F00000F0F0,
     & Z0F0F00000F0F0000, ZFF00FF0000FF00FF
        COMMON /LARGE_NUM2/ Z00FF00FF00000000, Z00000000FF00FF00,
     & Z3333333333333333, Z1111111111111111

        ROW_MASK = 65535
        T8_1 = 983055
        COL_MASK = IOR(ISHFT(T8_1, 32), T8_1)

        Z000F = 15
        Z00F0 = 240
        Z0F00 = 3840

        T8_1 = 61680
        T8_2 = 3855
        ZF0F00F0F = IOR(ISHFT(T8_1, 16), T8_2)
        T8_1 = 65280
        ZFF00FF00 = IOR(ISHFT(T8_1, 16), T8_1) 

        ZF0F00F0FF0F00F0F = IOR(ISHFT(ZF0F00F0F, 32), ZF0F00F0F)
        T8_1 = 61680
        Z0000F0F00000F0F0 = IOR(ISHFT(T8_1, 32), T8_1)
        T8_1 = 252641280
        Z0F0F00000F0F0000 = IOR(ISHFT(T8_1, 32), T8_1)
        T8_1 = 16711935
        ZFF00FF0000FF00FF = IOR(ISHFT(ZFF00FF00, 32), T8_1)
        T8_1 = 16711935
        Z00FF00FF00000000 = ISHFT(T8_1, 32)
        Z00000000FF00FF00 = ZFF00FF00
        T8_1 = 858993459
        Z3333333333333333 = IOR(ISHFT(T8_1, 32), T8_1)
        T8_1 = 286331153
        Z1111111111111111 = IOR(ISHFT(T8_1, 32), T8_1)
      END

      INTEGER*4 FUNCTION UNIF_RANDOM(N)
        INTEGER*4 :: N
        EXTERNAL c_rand
        INTEGER*4 :: c_rand

        UNIF_RANDOM = MOD(c_rand(), N)
      END

      INTEGER*8 FUNCTION UNPACK_COL(ROW)
        INTEGER*2 :: ROW
        INTEGER*8 :: T0, T1, T2, T3
        INTEGER*8 :: ROW_MASK, COL_MASK
        COMMON /MASK_NUM/ ROW_MASK, COL_MASK

        T0 = ROW
        T0 = IAND(T0, ROW_MASK)
        T1 = ISHFT(T0, 12)
        T2 = ISHFT(T0, 24)
        T3 = ISHFT(T0, 36)

        UNPACK_COL = IAND(IOR(IOR(IOR(T0, T1), T2), T3), COL_MASK)
      END

      INTEGER*2 FUNCTION REVERSE_ROW(ROW)
        INTEGER*2 :: ROW
        INTEGER*2 :: T0, T1, T2, T3
        INTEGER*2 :: Z000F, Z00F0, Z0F00
        COMMON /CONST_NUM/  Z000F, Z00F0, Z0F00

        T0 = ISHFT(ROW, -12)
        T1 = IAND(ISHFT(ROW, -4), Z00F0)
        T2 = IAND(ISHFT(ROW, 4), Z0F00)
        T3 = ISHFT(ROW, 12)

        REVERSE_ROW = IOR(IOR(IOR(T0, T1), T2), T3)
      END

      SUBROUTINE PRINT_BOARD(BOARD)
        INTEGER*8 :: BOARD
        INTEGER*8 :: BOARD_, T_
        INTEGER*4 :: I, J, POWER_VAL
        INTEGER*2 :: Z000F, Z00F0, Z0F00
        COMMON /CONST_NUM/  Z000F, Z00F0, Z0F00

        BOARD_ = BOARD
        T_ = Z000F
        WRITE(*, *) '-----------------------------'
        DO I = 0, 3
          DO J = 0, 3
            POWER_VAL = IAND(BOARD_, T_)
            IF (POWER_VAL == 0) THEN
              WRITE(*, '(A, "      ", $)') '|'
            ELSE
              WRITE(*, '(A, I6, $)')
     & '|', ISHFT(1, POWER_VAL)
            END IF
            BOARD_ = ISHFT(BOARD_, -4)
          END DO
          WRITE(*, '(A)') '|'
        END DO
        WRITE(*, *) '-----------------------------'
      END

      INTEGER*8 FUNCTION TRANSPOSE_BOARD(X)
        INTEGER*8 :: X
        INTEGER*8 :: A1, A2, A3, A, B1, B2, B3
        INTEGER*8 :: ZF0F00F0FF0F00F0F, Z0000F0F00000F0F0,
     & Z0F0F00000F0F0000, ZFF00FF0000FF00FF
        INTEGER*8 :: Z00FF00FF00000000, Z00000000FF00FF00,
     & Z3333333333333333, Z1111111111111111
        COMMON /LARGE_NUM1/ ZF0F00F0FF0F00F0F, Z0000F0F00000F0F0,
     & Z0F0F00000F0F0000, ZFF00FF0000FF00FF
        COMMON /LARGE_NUM2/ Z00FF00FF00000000, Z00000000FF00FF00,
     & Z3333333333333333, Z1111111111111111

        A1 = IAND(X, ZF0F00F0FF0F00F0F)
        A2 = IAND(X, Z0000F0F00000F0F0)
        A3 = IAND(X, Z0F0F00000F0F0000)
        A = IOR(IOR(A1, ISHFT(A2, 12)), ISHFT(A3, -12))
        B1 = IAND(A, ZFF00FF0000FF00FF)
        B2 = IAND(A, Z00FF00FF00000000)
        B3 = IAND(A, Z00000000FF00FF00)

        TRANSPOSE_BOARD = IOR(IOR(B1, ISHFT(B2, -24)), ISHFT(B3, 24))
      END FUNCTION TRANSPOSE_BOARD

      INTEGER*4 FUNCTION COUNT_EMPTY(X)
        INTEGER*8 :: X
        INTEGER*8 :: X_, T_
        INTEGER*2 :: Z000F, Z00F0, Z0F00
        COMMON /CONST_NUM/  Z000F, Z00F0, Z0F00
        INTEGER*8 :: Z00FF00FF00000000, Z00000000FF00FF00,
     & Z3333333333333333, Z1111111111111111
        COMMON /LARGE_NUM2/ Z00FF00FF00000000, Z00000000FF00FF00,
     & Z3333333333333333, Z1111111111111111

        X_ = X
        X_ = IOR(X_, IAND(ISHFT(X_, -2), Z3333333333333333))
        X_ = IOR(X_, ISHFT(X_, -1))
        X_ = IAND(NOT(X_), Z1111111111111111)
        X_ = X_ + ISHFT(X_, -32)
        X_ = X_ + ISHFT(X_, -16)
        X_ = X_ + ISHFT(X_, -8)
        X_ = X_ + ISHFT(X_, -4)
        T_ = Z000F
        COUNT_EMPTY = IAND(X_, T_)
      END

      INTEGER*2 FUNCTION EXECUTE_MOVE_HELPER(ROW)
        INTEGER*2 :: ROW
        INTEGER*2 :: I, J, T0, T1, T2, T3
        INTEGER*2 :: ROW_LINE(0:3)
        INTEGER*2 :: Z000F, Z00F0, Z0F00
        COMMON /CONST_NUM/  Z000F, Z00F0, Z0F00

        ROW_LINE(0) = IAND(ROW, Z000F)
        ROW_LINE(1) = IAND(ISHFT(ROW, -4), Z000F)
        ROW_LINE(2) = IAND(ISHFT(ROW, -8), Z000F)
        ROW_LINE(3) = IAND(ISHFT(ROW, -12), Z000F)
        I = 0
        DO WHILE (I < 3)
          J = I + 1
          DO WHILE (J < 4)
            IF (ROW_LINE(J) /= 0) THEN
              EXIT
            END IF
            J = J + 1
          END DO
          IF (J == 4) THEN
            EXIT
          END IF
          IF (ROW_LINE(I) == 0) THEN
            ROW_LINE(I) = ROW_LINE(J)
            ROW_LINE(J) = 0
            I = I - 1
          ELSE IF (ROW_LINE(I) == ROW_LINE(J)) THEN
            IF (ROW_LINE(I) /= 15) THEN
              ROW_LINE(I) = ROW_LINE(I) + 1
            END IF
            ROW_LINE(J) = 0
          END IF
          I = I + 1
        END DO
        T0 = ROW_LINE(0)
        T1 = ISHFT(ROW_LINE(1), 4)
        T2 = ISHFT(ROW_LINE(2), 8)
        T3 = ISHFT(ROW_LINE(3), 12)
        EXECUTE_MOVE_HELPER = IOR(IOR(IOR(T0, T1), T2), T3)
      END

      INTEGER*8 FUNCTION EXECUTE_MOVE(BOARD, MOVE)
        INTEGER*8 :: BOARD
        INTEGER*4 :: MOVE
        INTEGER*8 :: RET, T
        INTEGER*2 :: ROW, REV_ROW, I
        INTEGER*8 :: TRANSPOSE_BOARD, UNPACK_COL
        INTEGER*2 :: REVERSE_ROW, EXECUTE_MOVE_HELPER
        INTEGER*8 :: ROW_MASK, COL_MASK
        COMMON /MASK_NUM/ ROW_MASK, COL_MASK

        RET = BOARD
        IF ((MOVE == 0) .OR. (MOVE == 1)) THEN
            T = TRANSPOSE_BOARD(BOARD)
        ELSE
            T = BOARD
        END IF
        DO I = 0, 3
          ROW = IAND(ISHFT(T, -ISHFT(I, 4)), ROW_MASK)
          IF (MOVE == 0) THEN
            RET = IEOR(RET, ISHFT(UNPACK_COL(IEOR(ROW,
     & EXECUTE_MOVE_HELPER(ROW))), ISHFT(I, 2)))
          ELSE IF (MOVE == 1) THEN
            REV_ROW = REVERSE_ROW(ROW)
            RET = IEOR(RET, ISHFT(UNPACK_COL(IEOR(ROW, REVERSE_ROW(
     & EXECUTE_MOVE_HELPER(REV_ROW)))), ISHFT(I, 2)))
          ELSE IF (MOVE == 2) THEN
            RET = IEOR(RET, ISHFT(IAND(IEOR(ROW,
     & EXECUTE_MOVE_HELPER(ROW)), ROW_MASK), ISHFT(I, 4)))
          ELSE IF (MOVE == 3) THEN
            REV_ROW = REVERSE_ROW(ROW)
            RET = IEOR(RET, ISHFT(IAND(IEOR(ROW, REVERSE_ROW(
     & EXECUTE_MOVE_HELPER(REV_ROW))), ROW_MASK), ISHFT(I, 4)))
          END IF
        END DO
        EXECUTE_MOVE = RET
      END

      INTEGER*4 FUNCTION SCORE_HELPER(BOARD)
        INTEGER*8 :: BOARD
        INTEGER*4 :: SCORE, I, J, RANK
        INTEGER*2 :: ROW
        INTEGER*2 :: Z000F, Z00F0, Z0F00
        COMMON /CONST_NUM/  Z000F, Z00F0, Z0F00
        INTEGER*8 :: ROW_MASK, COL_MASK
        COMMON /MASK_NUM/ ROW_MASK, COL_MASK

        SCORE = 0
        DO J = 0, 3
          ROW = IAND(ISHFT(BOARD, -ISHFT(J, 4)), ROW_MASK)
          DO I = 0, 3
            RANK = IAND(ISHFT(ROW, -ISHFT(I, 2)), Z000F)
            IF (RANK >= 2) THEN
              SCORE = SCORE + ((RANK - 1) * ISHFT(1, RANK))
            END IF
          END DO
        END DO
        SCORE_HELPER = SCORE
      END

      INTEGER*4 FUNCTION SCORE_BOARD(BOARD)
        INTEGER*8 :: BOARD
        INTEGER*4 :: SCORE_HELPER

        SCORE_BOARD = SCORE_HELPER(BOARD)
      END

      INTEGER*8 FUNCTION DRAW_TILE()
        INTEGER*4 :: RET
        INTEGER*4 :: UNIF_RANDOM

        RET = UNIF_RANDOM(10)
        IF (RET < 9) THEN
          RET = 1
        ELSE
          RET = 2
        END IF
        DRAW_TILE = RET
      END

      INTEGER*8 FUNCTION INSERT_TILE_RAND(BOARD, TILE)
        INTEGER*8 :: BOARD
        INTEGER*8 :: TILE
        INTEGER*8 :: TILE_
        INTEGER*8 :: TMP, T_
        INTEGER*4 :: POS
        INTEGER*4 :: UNIF_RANDOM, COUNT_EMPTY
        INTEGER*2 :: Z000F, Z00F0, Z0F00
        COMMON /CONST_NUM/  Z000F, Z00F0, Z0F00

        TILE_ = TILE
        POS = UNIF_RANDOM(COUNT_EMPTY(BOARD))
        TMP = BOARD
        T_ = Z000F
        DO WHILE (1>0)
          DO WHILE (IAND(TMP, T_) /= 0)
            TMP = ISHFT(TMP, -4)
            TILE_ = ISHFT(TILE_, 4)
          END DO
          IF (POS == 0) THEN
            EXIT
          END IF
          POS = POS - 1
          TMP = ISHFT(TMP, -4)
          TILE_ = ISHFT(TILE_, 4)
        END DO
        INSERT_TILE_RAND = IOR(BOARD, TILE_)
      END

      INTEGER*8 FUNCTION INITIAL_BOARD()
        INTEGER*8 :: BOARD
        INTEGER*4 :: RD
        INTEGER*8 :: TILE
        INTEGER*4 :: UNIF_RANDOM
        INTEGER*8 :: DRAW_TILE, INSERT_TILE_RAND

        RD = UNIF_RANDOM(16)
        TILE = DRAW_TILE()
        BOARD = ISHFT(TILE, ISHFT(RD, 2))
        TILE = DRAW_TILE()
        INITIAL_BOARD = INSERT_TILE_RAND(BOARD, TILE)
      END

      INTEGER*4 FUNCTION ASK_FOR_MOVE(BOARD)
        INTEGER*8 :: BOARD
        INTEGER*4 :: RET
        CHARACTER :: MOVECHAR
        INTEGER :: POS
        CHARACTER(LEN=9) :: ALLMOVES = 'wsadkjhl'
        EXTERNAL c_getch
        INTEGER :: c_getch

        CALL PRINT_BOARD(BOARD)
        RET = -1
        DO WHILE (1 > 0)
          MOVECHAR = ACHAR(c_getch())
          IF (MOVECHAR == 'q') THEN
            RET = -1
            EXIT
          ELSE IF (MOVECHAR == 'r') THEN
            RET = 5
            EXIT
          END IF
          POS = INDEX(ALLMOVES, MOVECHAR)
          IF (POS /= 0) THEN
            RET = MOD(POS - 1, 4)
            EXIT
          END IF
        END DO
        ASK_FOR_MOVE = RET
      END

      SUBROUTINE PLAY_GAME()
        INTEGER*8 :: BOARD, NEWBOARD, TILE
        INTEGER*4 :: SCOREPENALTY, CURRENT_SCORE, LAST_SCORE, MOVENO
        INTEGER*8 :: RETRACT_VEC(0:63)
        INTEGER*1 :: RETRACT_PENALTY_VEC(0:63)
        INTEGER*4 :: RETRACT_POS, RETRACT_NUM, MOVE
        EXTERNAL c_clear_screen, c_print_move_score, c_print_final_score
        INTEGER*8 :: INITIAL_BOARD, EXECUTE_MOVE
        INTEGER*8 :: DRAW_TILE, INSERT_TILE_RAND
        INTEGER*4 :: SCORE_BOARD, ASK_FOR_MOVE

        BOARD = INITIAL_BOARD()
        SCOREPENALTY = 0
        CURRENT_SCORE = 0
        LAST_SCORE = 0
        MOVENO = 0
        RETRACT_POS = 0
        RETRACT_NUM = 0

        DO WHILE (1 > 0)
          CALL c_clear_screen()
          MOVE = 0
          DO WHILE (MOVE < 4)
            IF (EXECUTE_MOVE(BOARD, MOVE) /= BOARD) THEN
              EXIT
            END IF
            MOVE = MOVE + 1
          END DO
          IF (MOVE == 4) THEN
            EXIT
          END IF

          CURRENT_SCORE = SCORE_BOARD(BOARD) - SCOREPENALTY
          MOVENO = MOVENO + 1
          CALL c_print_move_score(MOVENO, CURRENT_SCORE, LAST_SCORE)
          LAST_SCORE = CURRENT_SCORE

          MOVE = ASK_FOR_MOVE(BOARD)
          IF (MOVE < 0) THEN
            EXIT
          END IF

          IF (MOVE == 5) THEN
            IF ((MOVENO <= 1) .OR. (RETRACT_NUM <= 0)) THEN
              MOVENO = MOVENO - 1
              CYCLE
            END IF
            MOVENO = MOVENO - 2
            IF ((RETRACT_POS == 0) .AND. (RETRACT_NUM > 0)) THEN
              RETRACT_POS = 64
            END IF
            RETRACT_POS = RETRACT_POS - 1
            BOARD = RETRACT_VEC(RETRACT_POS)
            SCOREPENALTY = 
     & SCOREPENALTY - RETRACT_PENALTY_VEC(RETRACT_POS)
            RETRACT_NUM = RETRACT_NUM - 1
            CYCLE
          END IF

          NEWBOARD = EXECUTE_MOVE(BOARD, MOVE)
          IF (NEWBOARD == BOARD) THEN
            MOVENO = MOVENO - 1
            CYCLE
          END IF

          TILE = DRAW_TILE()
          IF (TILE == 2) THEN
            SCOREPENALTY = SCOREPENALTY + 4
            RETRACT_PENALTY_VEC(RETRACT_POS) = 4
          ELSE
            RETRACT_PENALTY_VEC(RETRACT_POS) = 0
          END IF
          RETRACT_VEC(RETRACT_POS) = BOARD
          RETRACT_POS = RETRACT_POS + 1
          IF (RETRACT_POS == 64) THEN
            RETRACT_POS = 0
          END IF
          IF (RETRACT_NUM < 64) THEN
            RETRACT_NUM = RETRACT_NUM + 1
          END IF
          BOARD = INSERT_TILE_RAND(NEWBOARD, TILE)
        END DO
        CALL PRINT_BOARD(BOARD)
        CALL c_print_final_score(CURRENT_SCORE)
      END

      SUBROUTINE MAIN()
        CALL INIT_NUM()
        CALL PLAY_GAME()
      END

